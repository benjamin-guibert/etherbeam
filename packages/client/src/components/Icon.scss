@use '../styles/resources/sizes';

.my-icon {
  display: inline-flex;
  align-items: center;

  &-label {
    padding-left: sizes.$spacer-s;
  }
}
