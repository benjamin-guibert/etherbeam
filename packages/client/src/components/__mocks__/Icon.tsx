import React, { ReactElement } from 'react'

const Icon = (): ReactElement => {
  return <span className="my-icon"></span>
}

export default Icon
