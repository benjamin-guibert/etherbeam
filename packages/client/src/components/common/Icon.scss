@import 'styles/variables';

.app-icon {
  &-xs {
    width: $app-size-xs;
    height: $app-size-xs;
  }

  &-sm {
    width: $app-size-sm;
    height: $app-size-sm;
  }

  &-m {
    width: $app-size-m;
    height: $app-size-m;
  }

  &-lg {
    width: $app-size-lg;
    height: $app-size-lg;
    vertical-align: -25%;
  }

  &-max {
    width: 100%;
    height: 100%;
  }

  @for $i from 2 through 10 {
    &-#{$i} {
      width: #{$i}em;
      height: #{$i}em;
    }
  }
}
