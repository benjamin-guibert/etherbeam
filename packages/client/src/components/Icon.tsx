import React, { FC } from 'react'
import { IconProp } from '@fortawesome/fontawesome-svg-core'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import './Icon.scss'

type LabelBreakpointProp = 'xs' | 's' | 'm' | 'l'
type ColorProp = 'dark' | 'positive' | 'negative'
type SizeProp = 'm' | 'l' | 'xl'

interface IconProps {
  icon: IconProp
  label?: string
  labelBreakpoint?: LabelBreakpointProp
  size?: SizeProp
  iconColor?: ColorProp
  iconSize?: SizeProp
  className?: string
}

const Icon: FC<IconProps> = ({ icon, label, labelBreakpoint, size, iconColor, iconSize, className }) => {
  const classNames = ['my-icon', size ? `my-size-${size}` : '', className].join(' ')
  const iconClassName = [iconColor ? `my-icon-${iconColor}` : '', iconSize ? `my-size-${iconSize}` : ''].join(' ')
  const labelClassName = ['my-icon-label', labelBreakpoint ? `my-d-min-${labelBreakpoint}` : ''].join(' ')

  return (
    <span className={classNames}>
      <span className={iconClassName}>
        <FontAwesomeIcon icon={icon} />
      </span>
      {!!label && <span className={labelClassName}>{label}</span>}
    </span>
  )
}

export default Icon
