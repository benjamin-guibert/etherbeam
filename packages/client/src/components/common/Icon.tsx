import React, { ReactElement } from 'react'
import './Icon.scss'

interface IconProps {
  src: unknown
  alt?: string
  size?: 'xs' | 'sm' | 'm' | 'lg' | 'xl' | number
  className?: string
  noVerticalAlignment?: boolean
}

const Icon = ({ src, alt = '', size = 'm', className = '', noVerticalAlignment }: IconProps): ReactElement => {
  const getClasses = () => {
    const classes = ['app-icon', `app-icon-${size}`]

    if (noVerticalAlignment) classes.push('app-va-0')
    if (className) classes.push(className)

    return classes.join(' ')
  }

  return <img src={src as string} alt={alt} className={getClasses()} />
}

export default Icon
