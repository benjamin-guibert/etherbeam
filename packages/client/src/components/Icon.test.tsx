import React from 'react'
import { shallow, ShallowWrapper } from 'enzyme'
import Icon from './Icon'

jest.mock('@fortawesome/react-fontawesome', () => {
  return {
    __esModule: true,
    FontAwesomeIcon: () => <svg />,
  }
})

describe('<Icon />', () => {
  let component: ShallowWrapper

  describe('Default', () => {
    beforeAll(() => {
      component = shallow(<Icon label="Label" icon={null} />)
    })

    it('should render', () => {
      const span = component.find('span').at(0)
      expect(span.hasClass('my-icon')).toBeTruthy()
      const iconSpan = span.children().at(0)
      expect(iconSpan.hasClass('my-icon-svg')).toBeTruthy()
      expect(iconSpan.find('FontAwesomeIcon').exists()).toBeTruthy()
      const labelSpan = span.children().at(1)
      expect(labelSpan.hasClass('my-icon-label')).toBeTruthy()
    })

    it('should match snapshot', () => expect(component.render()).toMatchSnapshot())
  })

  describe('Breakpoint', () => {
    beforeAll(() => {
      component = shallow(<Icon label="Label" labelBreakpoint="m" icon={null} />)
    })

    it('should render', () => {
      const span = component.find('span').at(0).children().at(1)
      expect(span.hasClass('my-icon-label')).toBeTruthy()
      expect(span.hasClass('my-d-min-m')).toBeTruthy()
    })
  })

  describe('Class', () => {
    beforeAll(() => {
      component = shallow(<Icon label="Label" className="my-class" icon={null} />)
    })

    it('should render', () => {
      const span = component.find('span').at(0)
      expect(span.hasClass('my-icon')).toBeTruthy()
      expect(span.hasClass('my-class')).toBeTruthy()
    })
  })

  describe('No label', () => {
    beforeAll(() => {
      component = shallow(<Icon icon={null} />)
    })

    it('should render', () => expect(component.find('span').at(0).children()).toHaveLength(1))
  })
})
