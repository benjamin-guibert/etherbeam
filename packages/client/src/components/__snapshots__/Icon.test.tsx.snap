// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Icon /> Default should match snapshot 1`] = `
<span
  class="my-icon "
>
  <span
    class="my-icon-svg"
  >
    <svg />
  </span>
  <span
    class="my-icon-label"
  >
    Label
  </span>
</span>
`;
